### PR TITLE
Add workload pinning subsystem

### PR DESCRIPTION
**Pull request by MOZGIII**
This PR adds the workload pinning subsystem.

In comparison to the old implementation, within the new architecture this subsystem partially replaces the instance queue and locks - while not including any of the actual execution state (i.e. graphs in the old implementation or VM snapshots in the new implementation).

The workload can be anything, but in the current design it is intended to be used for managing the VM runtimes scheduling on the pool of executor nodes (in concrete terms - the `start-workers` instances).

The benefit of this approach is that it decouples the slow and costly operations on the big blobs from the more lightweight data that we can poll often (i.e. the queue and the locking status of a particular workload - or, more concretely - a instance of a VM).

---

This implementation expresses the interaction between the internal control loops in a tight type-safe manner, and also exposes underlying error conditions for explicit error handling.

---

The test coverage is good, and we explicitly cover tricky edge cases, like the behaviour of one loop when the other crashes.

### CODE REVIEW

File: crates/lib/workload-pinning-manager/src/maintenance/mod.rs [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+                        if !active_ids.is_empty() {
+                            info!("poll loop exited, {} active IDs remain", active_ids.len());
+                        }
+                    }
+                }
+            }
+            Some(id) = evict_rx.recv() => {
+                active_ids.remove(&id);
+                let _ = count_tx.send(active_ids.len());
+                if let Err(error) = (*backend).release_pinnings(node_id.clone(), NEVec::new(id)).await {
```
Comment (piercefreeman) [inline crates/lib/workload-pinning-manager/src/maintenance/mod.rs]: It might be worth a comment here that even if this fails, we'll still remove the id from active_ids. So effectively we rely on the TTL expiration to resume this work even during this error.

File: crates/lib/workload-pinning-manager/src/pinned_handle.rs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        Self {
+            id: Some(id),
+            evict_tx,
+        }
+    }
+
+    /// Return a reference to the wrapped instance ID.
+    pub fn id(&self) -> &InstanceId {
+        // SAFETY: `PinnedHandle` is not dropped, so the `id` must be present.
+        unsafe { self.id.as_ref().unwrap_unchecked() }
```
Comment (piercefreeman) [inline crates/lib/workload-pinning-manager/src/pinned_handle.rs:30]: Not totally clear why in the lifecycle this must be unsafe.

Comment (MOZGIII) [reply crates/lib/workload-pinning-manager/src/pinned_handle.rs:30]: `unsafe` block is required because `unwrap_unchecked` fn is unsafe. We don't *have to* use `unwrap_unchecked` here - but we can, and it's cheaper. The whole scope of safety concerns is pretty narrowly encapsulated in this tiny `PinnedHandle` type, so it's unlikely to cause issues during maintenance.

File: crates/lib/workload-pinning-manager/src/maintenance/mod.rs (hunk 1 above)
Comment (MOZGIII) [reply crates/lib/workload-pinning-manager/src/maintenance/mod.rs]: Actually, if we think about it, this is kind of bad; the backend error here will cause the overall loop to terminate - including the heartbeats. The parent loop cleanup will try to clean all the active ids before we exit - except this one.
So, if the cleanup succeeds, we'd be able to quickly re-pin the VMs in a new (or same) node - except for that one that was unlucky to hit the `release_pinnings` fault.

So yeah, this is a good find; also not covered by tests - will add it.

Comment (MOZGIII) [reply crates/lib/workload-pinning-manager/src/maintenance/mod.rs]: https://github.com/piercefreeman/waymark/compare/c85a02f2304efa9cf8be484f36df338b9b252ee3..ba5ca30aaa1547a15e67f0758f0e33b51b5fcab0

**Review by piercefreeman — APPROVED** [review]
New changes after review good to me!
